apis/v1beta1/shared_types.go

```diff
@@ -243,6 +243,40 @@ type PortNumber int32
 // ReferenceGrant object is required in the referent namespace to allow that
 // namespace's owner to accept the reference. See the ReferenceGrant
 // documentation for details.
+//
+// <gateway:experimental:description>
+//
+// When the BackendRef points to a Kubernetes Service implementations MUST allow
+// the backend protocol to be specified by:
+//   - setting the protocol field on the Service's ServicePort
+//   - setting the protocol field on the Service's related Endpoint/EndpointSlice's EndpointPort
+//   - setting the appProtocol field on the Service's ServicePort
+//   - setting the appProtocol field on the Service's related Endpoint/EndpointSlice's EndpointPort
```

Comment:
Note that at least in the case of Service, the values of the Protocol and AppProtocol fields are copied verbatim to Endpoints and EndpointSlices. Of course users could create custom EndpointSlices with different values but I don't think we want to honor or support that.

In general, I think this addition to the spec may be saying more than we need to. A lot of the content could potentially be replaced with something like this with the idea that implementers can read AppProtocol docs for more information:

```suggestion
// When a BackendRef points to a Kubernetes Service, implementations SHOULD
// honor the appProtocol field if it is set for the target Service Port.
```

I think that we're likely going to be stuck with "SHOULD" because I think AppProtocol has always been more of a "hint" than a requirement. I don't think we can change that retroactively.